lib/Transforms/HandshakeOptimization.cpp

```diff
@@ -0,0 +1,629 @@
+    if (it != cmergeOp->getUsers().end()) {
+      useOwner = *it;
+    }
+    if (isa_and_nonnull<handshake::ConditionalBranchOp>(useOwner)) {
+      handshake::ConditionalBranchOp condBranchOp =
+          mlir::cast<handshake::ConditionalBranchOp>(useOwner);
+
+      mlir::OperandRange dataOperands = cmergeOp.getDataOperands();
+      mlir::Value first = dataOperands[0];
+      mlir::Value second = dataOperands[1];
+      mlir::Value resultCMerge = cmergeOp.getResult();
+      mlir::Value index = cmergeOp.getIndex();
+      mlir::Value condition = condBranchOp.getConditionOperand();
+      mlir::Value branchIn = condBranchOp.getDataOperand();
+      mlir::Value trueResult = condBranchOp.getTrueResult();
+      mlir::Value falseResult = condBranchOp.getFalseResult();
+
+      if (resultCMerge == branchIn) {
+        mlir::Type constantType = rewriter.getIntegerType(1);
+        mlir::Value zero = rewriter.create<handshake::ConstantOp>(
+            cmergeOp->getLoc(), constantType,
+            rewriter.getIntegerAttr(constantType, 0), start);
+        mlir::Value one = rewriter.create<handshake::ConstantOp>(
+            cmergeOp->getLoc(), constantType,
+            rewriter.getIntegerAttr(constantType, 1), start);
+        handshake::NotOp notOp =
+            rewriter.create<handshake::NotOp>(cmergeOp->getLoc(), condition);
+        mlir::Value valueOfConstant;
+        mlir::Value inputToMerge;
+        if (trueResult == first) {
+          valueOfConstant = one;
+          inputToMerge = notOp.getResult();
+          rewriter.replaceAllUsesWith(falseResult, second);
+          rewriter.replaceAllUsesWith(resultCMerge, second);
+        } else if (falseResult == second) {
+          valueOfConstant = zero;
+          inputToMerge = notOp.getResult();;
+          rewriter.replaceAllUsesWith(trueResult, first);
+          rewriter.replaceAllUsesWith(resultCMerge, first);
+        } else if (trueResult == second) {
+          valueOfConstant = zero;
+          inputToMerge = condition;
+          rewriter.replaceAllUsesWith(falseResult, first);
+          rewriter.replaceAllUsesWith(resultCMerge, first);
+        } else if (falseResult == first) {
+          valueOfConstant = one;
+          inputToMerge = condition;
+          rewriter.replaceAllUsesWith(trueResult, second);
+          rewriter.replaceAllUsesWith(resultCMerge, second);
+        } else {
+          return failure();
+        }
+        ValueRange operands = {inputToMerge, valueOfConstant};
+        rewriter.setInsertionPointAfter(cmergeOp);
+        handshake::MergeOp mergeOp =
+            rewriter.create<handshake::MergeOp>(cmergeOp.getLoc(), operands);
+        Value mergeResult = mergeOp.getResult();
+        rewriter.replaceAllUsesWith(index, mergeResult);
+        rewriter.eraseOp(cmergeOp);
+        rewriter.eraseOp(condBranchOp);
+        return success();
+      }
+    }
+    return failure();
+  }
+};
+
+struct RemoveConsecutiveForksPairs
```

Comment:
If I trust the pattern's name then this optimization is already implemented [as part of the materialization pass](https://github.com/EPFL-LAP/dynamatic/blob/682d5efcfb127e10cbcdca48c134ebd875133945/lib/Transforms/HandshakeMaterialize.cpp#L230-L273).